contracts/LockRelease.sol

```diff
@@ -1,101 +1,162 @@
+    /** Deploys the LockRelease contract and sets up all beneficiary release schedules.
+     *
+     * @param _token address of the beneficiary
+     * @param _beneficiaries array of beneficiary addresses to create release schedules for
+     * @param _amounts array of the amount of tokens to be locked and released for each beneficiary
+     * @param _start the timestamp that the release schedule begins releasing tokens at
+     * @param _duration the time period in seconds that tokens are released over
+     */
+    constructor(
+        address _token,
+        address[] memory _beneficiaries,
+        uint256[] memory _amounts,
+        uint128 _start,
+        uint128 _duration
+    ) EIP712("DecentLockRelease", "1.0.0") {
+        if (_token == address(0)) revert InvalidToken();
+        if (_duration == 0) revert ZeroDuration();
+        if (_beneficiaries.length != _amounts.length)
+            revert InvalidArrayLengths();
+
+        for (uint16 i = 0; i < _beneficiaries.length; ) {
             uint256 amount = _amounts[i];
             if (amount == 0) revert InvalidAmount();
 
-            total += amount;
-
             address beneficiary = _beneficiaries[i];
             if (beneficiary == address(0)) revert InvalidBeneficiary();
-            if (schedules[beneficiary].total != 0) revert DuplicateBeneficiary();
+            if (schedules[beneficiary].total != 0)
+                revert DuplicateBeneficiary();
+
+            schedules[beneficiary] = Schedule(amount, 0);
 
-            Schedule memory schedule = Schedule(amount, 0);
-            schedules[beneficiary] = schedule;
+            // mint the beneficiary voting units
+            _transferVotingUnits(address(0), beneficiary, amount);
+
+            // beneficiary delegates to themselves
+            _delegate(beneficiary, beneficiary);
+
+            unchecked {
+                ++i;
+            }
         }
 
+        token = _token;
+        start = _start;
+        duration = _duration;
         beneficiaries = _beneficiaries;
 
-        // Transfer tokens from sender to contract
-        IERC20(token).safeTransferFrom(msg.sender, address(this), total);
+        emit ScheduleStarted(
+            _token,
+            _beneficiaries,
+            _amounts,
+            _start,
+            _duration
+        );
+    }
+
+    /** Release all releasable tokens to the caller. */
+    function release() external {
+        uint256 releasable = getReleasable(msg.sender);
+
+        if (releasable == 0) revert NothingToRelease();
+
+        // Update released amount
+        schedules[msg.sender].released =
+            schedules[msg.sender].released +
+            releasable;
 
-        emit ScheduleStarted(total, _start, _duration);
+        // Burn the released voting units
+        _transferVotingUnits(msg.sender, address(0), releasable);
+
+        // Transfer tokens to recipient
+        IERC20(token).safeTransfer(msg.sender, releasable);
```

Comment:
does it save gas to store `token` as an `IERC20` rather than address?

Reply:
Nope, same amount of gas, since behind the scenes the EVM is still just storing an address